notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/TagEnricher.java

```diff
@@ -75,7 +88,9 @@ public boolean removeTag(String tag) {
      */
     @Override
     public boolean removeTags(Collection<? extends String> tags) {
-        return mTags.removeAll(tags);
+        Set<String> set = (Set<String>) getTags();
+        set.removeAll(tags);
+        return addTags(set);

```

Comment:
A couple of things here: 
1. I'm not a fan of down-casting, even when it's on an object that we produced.
1.  We need to be careful of a weird behavior of `SharedPreferences.getStringSet` that's documented [here](https://developer.android.com/reference/android/content/SharedPreferences.html#getStringSet(java.lang.String,%20java.util.Set%3Cjava.lang.String%3E)):
    > Note that you must not modify the set instance returned by this call. The consistency of the stored data is not guaranteed if you do, nor is your ability to modify the instance at all.
    
    The net effect as noted in a few places is that the corresponding `putStringSet` call must be done with a new instance of a `Set<String>`.

The way I'd work around this weirdness:

``` Java
@Override
public boolean removeTags(Collection<? extends String> tags) {
    HashSet<String> toModify = new HashSet<String>(mSharedPreferences.getStringSet("tags", new HashSet<String>()));
    if (toModify.removeAll(tags)) {
        mSharedPrefrences.edit().putStringSet("tags", toModify).apply();
        return true;
    }
    return false;
}
```

My implementation for `addTags` would be very similar. :)